plugins/catalog-backend/src/next/DefaultCatalogProcessingEngine.ts

```diff
@@ -69,51 +72,60 @@ export class DefaultCatalogProcessingEngine implements CatalogProcessingEngine {
     for (const provider of this.entityProviders) {
       await provider.connect(
         new Connection({
-          stateManager: this.stateManager,
+          processingDatabase: this.processingDatabase,
           id: provider.getProviderName(),
         }),
       );
     }
-
     this.running = true;
+    this.run();
+  }
 
+  private async run() {
     while (this.running) {
-      const {
-        id,
-        entity,
-        state: initialState,
-      } = await this.stateManager.getNextProcessingItem();
-
-      const result = await this.orchestrator.process({
-        entity,
-        state: initialState,
+      const { items } = await this.processingDatabase.transaction(async tx => {
+        return this.processingDatabase.getProcessableEntities(tx, {
+          processBatchSize: 1,
+        });
       });
 
-      for (const error of result.errors) {
-        this.logger.warn(error.message);
-      }
-
-      if (!result.ok) {
-        return;
+      if (items.length) {
+        const { id, state, unprocessedEntity } = items[0];
+
+        const result = await this.orchestrator.process({
+          entity: unprocessedEntity,
+          state,
+        });
+        for (const error of result.errors) {
+          this.logger.warn(error.message);
+        }
+        if (!result.ok) {
+          return;
```

Comment:
should be continue, not return, right? (another reason to go for the wrapping in a separate function as suggested above)